src/consensus.rs

```diff
@@ -679,114 +941,120 @@ impl Consensus {
-                            err
-                        )
+                    None if self.bft_round_state.slot == 0 => {
+                        if consensus_proposal.slot != 0 {
+                            warn!("🔄 Consensus state out of sync, need to catchup");
+                        } else {
+                            info!("#### Received genesis block proposal ####");
+                            self.genesis_bond(&consensus_proposal.validators)?;
+                        }
+                    }
+                    None => {
+                        // Verify proposal hash is correct
+                        if !self.verify_consensus_proposal_hash(hash) {
+                            self.metrics.prepare_error("previous_commit_qc_unknown");
+                            // TODO Retrieve that data on other validators
+                            bail!("Previsou Commit Quorum certificate is about an unknown consensus proposal. Can't process any furter");
+                        }
+
+                        let previous_commit_quorum_certificate_with_message = SignedWithKey {
+                            msg: ConsensusNetMessage::ConfirmAck(hash.clone()),
+                            signature: cert.signature.clone(),
+                            validators: cert.validators.clone(),
+                        };
+                        // Verify valid signature
+                        match BlstCrypto::verify(&previous_commit_quorum_certificate_with_message) {
+                            Ok(res) => {
+                                if !res {
+                                    self.metrics.prepare_error("previous_commit_qc_invalid");
+                                    bail!("Previous Commit Quorum Certificate is invalid")
+                                }
+                                // Buffers the previous *Commit* Quorum Cerficiate
+                                self.store
+                                    .bft_round_state
+                                    .commit_quorum_certificates
+                                    .insert(
+                                        self.store.bft_round_state.slot - 1,
+                                        (hash.clone(), cert.clone()),
+                                    );
+                            }
+                            Err(err) => {
+                                self.metrics.prepare_error("bls_failure");
+                                bail!(
+                                    "Previous Commit Quorum Certificate verification failed: {}",
+                                    err
+                                )
+                            }
+                        }
+                        // Properly finish the previous round
+                        self.finish_round()?;
                     }
                 }
-                // Properly finish the previous round
-                self.finish_round()?;
+                
```

Comment:
Toute cette fonction me parait weird à vrai dire. Je comprends pas pourquoi on récupère le previous certificate au dessus, je comprends pas ce qu'on fait si on l'a pas.

Pour moi la logique là c'est qu'on process le CommitCertificate si pas déjà fait, et on a juste besoin dans Prepare de check que le Ticket a déjà été validé.

Je sais pas si c'est "in scope" de le réécrire mais je trouve le truc confusing.